harmony_py/auth.py

```diff
@@ -0,0 +1,146 @@
+from requests import Session
+from requests.models import PreparedRequest, Response
+from requests.utils import get_netrc_auth
+from requests_futures.sessions import FuturesSession
+from typing import cast, Optional, Tuple, Union
+
+from .config import Config
+
+
+cfg = Config()
+
+
+def _is_edl_hostname(hostname: str) -> bool:
+    """
+    Determine if a hostname matches an EDL hostname.
+
+    Parameters:
+        hostname (str): A fully-qualified domain name (FQDN).
+
+    Returns:
+        (boolean): True if the hostname is an EDL hostname, else False.
+    """
+    edl_hostname_pattern = r'.*urs\.earthdata\.nasa\.gov$'
+    return re.fullmatch(edl_hostname_pattern, hostname, flags=re.IGNORECASE) is not None
+
+
+class MalformedCredentials(Exception):
+    pass
+
+
+class BadAuthentication(Exception):
+    pass
+
+
+class SessionWithHeaderRedirection(Session):
+    """Modify Authorization headers in accordance with Earthdata Login (EDL) common usage.
+
+    Example:
+        session = SessionWithHeaderRedirection(username, password)
+
+    Parameters:
+        auth (Tuple(str, str)): A tuple of the format ('edl_username', 'edl_password')
+    """
+
+    def __init__(self, auth: Optional[Union[Tuple[str, str], None]] = None) -> None:
+        super().__init__()
+        if auth:
+            self.auth = auth
+        else:
+            self.auth = None
+
+    def rebuild_auth(self, prepared_request: PreparedRequest, response: Response) -> None:
+        """
+        Override Session.rebuild_auth. Strips the Authorization header if neither
+        original URL nor redirected URL belong to an Earthdata Login (EDL) host. Also
+        allows the default requests behavior of searching for relevant .netrc
+        credentials if and only if a username and password weren't provided during
+        object instantiation.
+
+        Parameters:
+            prepared_request (:obj:`PreparedRequest`): Object for the redirection
+            destination.
+            response (:obj:`PreparedRequest`): Object for the where we just came from.
+
+        Returns:
+            (boolean): True if the hostname is an EDL hostname, else False.
+        """
+
+        headers = prepared_request.headers
+        redirect_hostname = cast(str, urlparse(prepared_request.url).hostname)
+        original_hostname = cast(str, urlparse(response.request.url).hostname)
+
+        if 'Authorization' in headers \
+                and (original_hostname != redirect_hostname) \
+                and not _is_edl_hostname(redirect_hostname):
+            del headers['Authorization']
+
+        if self.auth is None:
+            # .netrc might have more auth for us on our new host.
+            new_auth = get_netrc_auth(prepared_request.url) if self.trust_env else None
+            if new_auth is not None:
+                prepared_request.prepare_auth(new_auth)
+
+        return
+
+
+def authenticate(*, auth: Optional[Union[Tuple[str, str], None]] = None,
```

Comment:
This is new syntax to me.  What is `*` as the first param?

Also, did we end up with:
`authenticate(auth=(username, password))`

Which could also be shortened to:
`authenticate((username, password))`

I'm curious about the reasoning.

Reply:
Essentially it forces the remaining arguments to be keyword-only. 

Using a keyword to pass the tuple we thought was preferable since we will add the ability to pass a token later. It communicates in the type signature that a tuple (username,password) can be passed, and (later) that a token can be passed. This is also how requests works (`auth=(username,pwd)`). Maybe instead of `auth` this should be named `creds` or something similar.

By the way I think this type signature is wrong. Shouldn't it be `Optional[Tuple[str, str]] = None`? 

Later when we add a token I think it would mean adding: `token: Optional[str]`.

In the end any of these will be valid (let's pretend we change `auth` to `creds`):

```
authenticate()  # envvars or netrc file
authenticate(creds=('kbeam', 'secret'))
authenticate(token='xyzzy')
```

Reply:
Changing auth to creds.

Reply:
> By the way I think this type signature is wrong. Shouldn't it be Optional[Tuple[str, str]] = None?

I swear I had that before and something complained. It's working now though - changed.